### PR TITLE
Controller Mapping / Provider validations part 2

### PR DESCRIPTION
**Pull request by SimonDarksideJ**
# XRTK - Mixed Reality Toolkit Change Request

## Overview

Extension of the Controller Mapping validations, as data providers also need validating that they exist when you add a new platform.

### CODE REVIEW

File: XRTK-Core/Packages/com.xrtk.core/Extensions/ArrayExtensions.cs [hunk 1]
```diff
@@ -29,5 +29,20 @@ public static bool IsValidArray(this Array array)
...
         }
+
+        /// <summary>
+        /// Extends an existing array to add a new item
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="array">The array to extend</param>
+        /// <param name="newItem">The item to add to the array</param>
+        /// <returns></returns>
+        public static T[] AddItem<T>(this T[] array, T newItem)
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Extensions/ArrayExtensions.cs]: optional index specification?

might be better if pulled into a separate pr?

Comment (SimonDarksideJ) [reply XRTK-Core/Packages/com.xrtk.core/Extensions/ArrayExtensions.cs]: Created #453.

Array.Copy (at least under Unity requires an index)

File: XRTK-Core/Packages/com.xrtk.core/Utilities/ValidateConfiguration.cs [hunk 2]
```diff
@@ -158,5 +179,73 @@ public static bool ValidateControllerProfiles(Type[] mappingTypesToValidate)
...
             return false;
         }
+
+        /// <summary>
+        /// Controller Mapping function to test for a controller mappings for a specific hand
+        /// </summary>
+        /// <param name="mappingTypesToValidate">Array of controller mappings to validate</param>
+        /// <param name="controllerHandedness">The <see cref="Handedness"/> of the controller to validate</param>
+        /// <returns></returns>
+        public static bool ValidateControllerMappings(Type[] mappingTypesToValidate, Handedness controllerHandedness)
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Utilities/ValidateConfiguration.cs]: Couldn't this just be an override?

It seems a lot of this is duplicated from above.

Comment (SimonDarksideJ) [reply XRTK-Core/Packages/com.xrtk.core/Utilities/ValidateConfiguration.cs]: No, because one looks at data providers, one looks at mapping profiles and the other looks at mapping interactions.

**Comment by SimonDarksideJ** [comment]
It's consumed in the respective platforms.  No point in validating config in core as it's provided in the defaults.  Unless you feel we should also apply this for OpenVR?

**Comment by StephenHodgson** [comment]
Can we add some unit tests then to verify these?

**Comment by SimonDarksideJ** [comment]
That is a fair point, Unit Tests incoming. Thanks for pulling me up on that (reserves that for future use in offence :D )

**Comment by StephenHodgson** [comment]
@SimonDarksideJ gotta hold ourselves accountable somehow, lol. I'm mainly just curious cause we don't have a working example in the codebase.

**Comment by SimonDarksideJ** [comment]
Added Unit tests where possible and made alterations to make the functions testable.
However, can't create controller mapping profiles programatically, so those can't be unit tested.

File: XRTK-Core/Assets/XRTK.Tests/Core/TestFixture_02_MixedRealityToolkitUtilityTests.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (StephenHodgson) [inline XRTK-Core/Assets/XRTK.Tests/Core/TestFixture_02_MixedRealityToolkitUtilityTests.cs]: I'm dubious of the licensing again. Is this copied from the MRTK?

File: XRTK-Core/Assets/XRTK.Tests/Core/TestFixture_02_MixedRealityToolkitUtilityTests.cs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            TestUtilities.InitializeMixedRealityToolkitScene();
+
+            InitializeDefaultInputSystemProfile();
+
+            Assert.IsFalse(ValidateConfiguration.ValidateControllerProfiles(controllerTypes, false));
+        }
+
+        private void InitializeDefaultInputSystemProfile()
+        {
+            var inputSystemProfile = new MixedRealityInputSystemProfile();
```
Comment (StephenHodgson) [inline XRTK-Core/Assets/XRTK.Tests/Core/TestFixture_02_MixedRealityToolkitUtilityTests.cs]: You can't use UnityEngine types like this

File: XRTK-Core/Packages/com.xrtk.core/Utilities/ValidateConfiguration.cs [hunk 5]
```diff
@@ -100,59 +128,139 @@ public static bool ValidateControllerProfiles(Type[] mappingTypesToValidate)
             if (MixedRealityToolkit.HasActiveProfile)
             {
                 var errorsFound = false;
-                var mappingConfigurationSource = MixedRealityToolkit.Instance.ActiveProfile.InputSystemProfile.ControllerMappingProfiles.MixedRealityControllerMappings;
+                var mappingConfigurationSource = MixedRealityToolkit.Instance.ActiveProfile.InputSystemProfile?.ControllerMappingProfiles?.MixedRealityControllerMappings;
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Utilities/ValidateConfiguration.cs]: Same thing there the `?.` operator doesn't work on `UnityEngine` types

File: XRTK-Core/Assets/XRTK.Tests/Core/TestFixture_02_MixedRealityToolkitUtilityTests.cs (hunk 4 above)
Comment (StephenHodgson) [reply XRTK-Core/Assets/XRTK.Tests/Core/TestFixture_02_MixedRealityToolkitUtilityTests.cs]: should have been:
```
var inputSystemProfile = ScriptableObject.CreateInstance<MixedRealityInputSystemProfile>();
            inputSystemProfile.ControllerDataProvidersProfile = ScriptableObject.CreateInstance<MixedRealityControllerDataProvidersProfile>();
            inputSystemProfile.ControllerMappingProfiles = ScriptableObject.CreateInstance<MixedRealityControllerMappingProfiles>();
            ```

Comment (StephenHodgson) [reply XRTK-Core/Assets/XRTK.Tests/Core/TestFixture_02_MixedRealityToolkitUtilityTests.cs]: https://github.com/JetBrains/resharper-unity/wiki/ScriptableObjects-must-be-instantiated-with-ScriptableObject.CreateInstance-instead-of-new

Comment (StephenHodgson) [reply XRTK-Core/Assets/XRTK.Tests/Core/TestFixture_02_MixedRealityToolkitUtilityTests.cs]: This will matter when we start doing runtime unit tests
